cdc/processor/pipeline/sink.go

```diff
@@ -215,6 +177,75 @@ func (n *sinkNode) emitEvent(ctx pipeline.NodeContext, event *model.PolymorphicE
 	return nil
 }
 
+// shouldSplitUpdateEvent determines if the split event is needed to align the old format based on
+// whether the handle key column has been modified.
+// If the handle key column is modified,
+// we need to use splitUpdateEvent to split the update event into a delete and an insert event.
+func shouldSplitUpdateEvent(updateEvent *model.PolymorphicEvent) bool {
+	// nil event will never be split.
+	if updateEvent == nil {
+		return false
+	}
+
+	handleKeyCount := 0
+	equivalentHandleKeyCount := 0
+	for i := range updateEvent.Row.Columns {
+		if updateEvent.Row.Columns[i].Flag.IsHandleKey() && updateEvent.Row.PreColumns[i].Flag.IsHandleKey() {
+			handleKeyCount++
+			colValueString := model.ColumnValueString(updateEvent.Row.Columns[i].Value)
+			preColValueString := model.ColumnValueString(updateEvent.Row.PreColumns[i].Value)
+			if colValueString == preColValueString {
+				equivalentHandleKeyCount++
+			}
+		}
+	}
+
+	// If the handle key columns are not updated, so we do **not** need to split the event row.
+	return !(handleKeyCount == equivalentHandleKeyCount)
+}
+
+// splitUpdateEvent splits an update event into a delete and an insert event.
+func splitUpdateEvent(updateEvent *model.PolymorphicEvent) ([]*model.PolymorphicEvent, error) {
```

Comment:
Or maybe we can retrun an array of 2 model.PolymorphicEvent to save alloctions.
```suggestion
func splitUpdateEvent(updateEvent *model.PolymorphicEvent) ([2]*model.PolymorphicEvent, error) {
```

Reply:
Solved by your previous comment.

Reply:
Returning an array is better than pre-allocation, as it does not need alloction in heap, see more: https://godbolt.org/z/K1Mdde5Kd

(I should mark https://github.com/pingcap/ticdc/pull/2367#discussion_r676615140 outdated before making this comment. 😅)

Reply:
Yes, but does this perhaps affect the readability of the code? Because we need to access it via index both for writing and reading.

```diff
-n.eventBuffer = append(n.eventBuffer, splitEvents...)
+n.eventBuffer = append(n.eventBuffer, splitEvents[0],splitEvents[1]) or n.eventBuffer = append(n.eventBuffer, splitEvents[:]...)
-result = append(result, &deleteEvent)
+result[0] = &deleteEvent
```

Also, if updateEvent is nil, what does our parameter detection need to return? 
```go
	var result [2]*model.PolymorphicEvent
	if updateEvent == nil {
		return result, errors.New("nil event cannot be split")
	}

```
Wouldn't that be a bit strange? Maybe we can keep it to improve code readability?

Reply:
return `[2]*model.PolymorphicEvent` also kind of strange to me 🤓

Reply:
I prefer `func splitUpdateEvent(updateEvent *model.PolymorphicEvent) (delEv, insertIv, *model.PolymorphicEvent, error)`

Reply:
Cool! Let me address that.

Reply:
Addressed. Thank you all for the review!